### PR TITLE
system76-scheduler 2.0.0

### LINKED ISSUE #55: Setting background and foreground to None in config.ron causes explicit process assignments to be ignored
Setting your config.ron to have
```rust
{
  background: None,
  foreground: None,
}
```
Causes any explicit process assignments (`assignments.ron` and others) to be not applied.

Seems to be a regression of #12, in which it was stated the desired behavior was to still allow per-process assignments to apply even with `None`.

If only `background` is `None`, then per-process mappings are only applied when a process goes into the foreground.
The opposite for `foreground`.

### LINKED ISSUE #69: CPU_PRIORITY is a misnomer
What you specify this way is the nice value, not `SHED_FIFO`/`SHED_RR` priority.

But actually it would be neat to have the ability to venture beyond the `SCHED_OTHER` policy: like, say, assign `SCHED_IDLE` to compilers etc. instead of assuming niceness 19 plus idle i/o is low enough for everybody.  Not sure about `SHED_RR`/`SHED_FIFO`, since those come with foot-guns attached, but maybe allow even those (not in the default config though).  I wonder if this direction has ever been explored in this project?

### LINKED ISSUE #70: ron package is out of date
The are some update on the branch your are pointing to including an update the version and deps

### LINKED ISSUE #73: Concept decent; execution needs work
After wrangling with PopOS for several months, I finally figured out that it was this daemon that was causing issues. Don't get me wrong - I like the concept. The _execution_, however, causes me major pain. In no particular order:

1. I was unable to see _why_ process priorities were being changed. I didn't spy anything in the logs, nor did I spy anything obvious in the system daemons either (because when I see 'scheduler daemon' I think of 'figuring out ordering/timing/priority of starting _new_ background tasks I own', not 'managing priorities of existing random other tasks not owned by me' - admittedly far more obvious in hindsight).
    * At least log something (rate-limited, of course.)
    * Looking at the logs with the benefit of hindsight, it does log something...
        * ...on boot only...
        * ...into the _system_ log not the _user_ log...
        * ...and the log message implies that it's only setting the priority of a specific few process names, not everything.
2. `nice` / `renice` just flat-out doesn't work. Oh, it seems to work... for a few seconds. Then reverts back to the old priority.
    * Solution: if you're going to fiddle with priorities, please make sure that if a priority is manually changed that it won't be fiddled with again.
        * Please also ensure that this handles the case of "someone switches the priority of a background process to a priority that happens to match the priority you assign foreground processes" (and vice versa.)
        * Please also ensure this doesn't fiddle with priorities of newly spawned subprocesses of said `renice`'d task.
3. Speaking of `nice`/`renice`, having the default foreground priority be lower (`-5`) than the minimum user-accessible priority (`0`) results in a lot of pain. 
    * Looking at the config file now, I note that you can change the configuration from `-5`. However, see point 1, above.
    * I'd suggest putting the defaults at `15` / `5` instead of `5` / `-5`. That way a user can launch a process at a higher priority than the default priority.
4. In general, there's no easy way to temporarily override the daemon's decision for a task, or to get visibility into why the daemon is making changes.
5. Priority inversions. So many priority inversions.
    * Fixing this requires a more capable configuration system that can express dependencies between processes, and a fair bit of work to actually put said dependencies into the configuration.
6. "The" foreground task is a poor proxy for what I actually care about.
    * My usual style of 'several terminals on top running things most of the time, and then flipping to Firefox for a few seconds before flipping back' unfortunately in practice often results in noticeable jank when switching from a terminal pegging the cpu to Firefox.
        * I think part of this is the following chain: background -> low priority -> process not running very often at all -> process gets mostly swapped out -> hiccough when window refocused. (Linux doesn't explicitly swap out low-priority processes, but the way kswapd interacts with process scheduling comes out to the same thing.)
        * Maybe prioritize all visible windows instead of simply 'the' currently focused window?
        * Maybe have a heuristic to keep processes that have been visible e.g. 10% of the last ten minutes foreground priority? (Or 'just' put recently-foreground'd processes in a cgroup with `vm.swappiness=0`? If you do this, make sure that if the process is already in a cgroup you make a nested cgroup, not clobber the existing one.)
    * In general, this style of daemon doesn't play well with anything that is _conditionally_ interactive (e.g. if I'm playing audio via Firefox, I want it to remain at a decent priority regardless of it it's in the background - but I don't _always_ want Firefox to remain at a boosted priority.)
        *  Other examples are `java` or `python` or `lua`. Should `java` be prioritized even if in the background? Answer: it depends entirely on what it's actually running.
        * Some of this may be solvable (with some false positives / false negatives) via e.g. hooking into the audio system to see what is currently using audio sources/sinks.

All told, on my system, with the sorts of workloads I run (as I type this I've got the cpu pegged with two separate 8-core bsat solver instances, for instance, both set to nice=19), system76-scheduler actually makes responsiveness significantly _worse_ than default.

(For reference: this is a Oryx Pro (9, not 10) with 32GB RAM.)

For now, I've `systemctl stop com.system76.Scheduler.service` / `systemctl disable com.system76.Scheduler.service` / `sudo apt remove system76-scheduler`'d it (and will continue to do so on all machines I control as necessary); I will revisit this in future if the worst of the warts are removed.

I figured you might like feedback as to _why_ I disabled it, as opposed to just 'one less user; why?  ¯\\\_(ツ)\_/¯'.

### LINKED ISSUE #74: Feature Request: allow disabling CFS tweaks
It would be great to allow disabling CFS tweaks, as it would help when it's already handled manually, or more simply when one doesn't want to change the default which comes with its kernel. 

A `use_cfstweaks: true|false` in the `config.ron` would be perfect

### LINKED ISSUE #76: Feature Request: configure processmonitor sleep time
https://github.com/pop-os/system76-scheduler/blob/0fe4d8dfc4275fd856aee28ca942b9fa53229fc9/daemon/src/main.rs#L310
```
tokio::time::sleep(Duration::from_secs(60)).await;
```
I often hangon with my mates on call and use easyeffects for my mic. It's not agile enough to manage priorities before i join the call. So i'd prefer to have this value down to 1sec configurable.

### PR DESCRIPTION
**Pull request by mmstick**
- Pipewire integration, which auto-assigns and retains a higher priority profile to PIDs connected to Pipewire
- Many performance optimizations
    - Eliminates most memory allocation by reusing buffers when reading files or building paths
    - Avoids performing UTF-8 string checks when not necessary, by using the `bstr`, `atoi`, and `bytelines` crates.
- Disables the kernel's autogroup feature by default so that nice priorities function properly
- Migrated from RON to KDL for configuration files
- Configuration files now have very flexible rules with many new options
- Now supports setting scheduler policies like SCHED_FIFO and SCHED_IDLE
- Assignments can be made based on conditions such as cgroups or parent processs
- Default configuration now automatically detects most system services without needing to explicitly name them
- The CFS profile service can be disabled in the config
- The process scheduler service can also be disabled
- The refresh rate of the process scheduler's PID monitoring service can be configured
- Foreground process assignments can be disabled by removing the `foreground` and/or `background` profile from the configuration file
- Pipewire integration can also be disabled by removing the `pipewire` profile assignment

Closes #76 
Closes #74 
Closes #73 
Closes #70 
Closes #69 
Closes #55

### CODE REVIEW

**Comment by 13r0ck** [comment]
This is really cool

File: README.md [hunk 1]
```diff
@@ -15,106 +15,73 @@ The `SetForeground(u32)` method can be called to change the active foreground pr
...
-```rs
-{
-    // The priority to assign background tasks.
-    background: Some(5),
-
-    // The priority to assign foreground tasks.
-    foreground: Some(-5),
-}
-```
+Presence of the system configuration will override the distribution configuration. The documented [default configuration can be found at here](./data/config.kdl).
```
Comment (jacobgkau) [inline README.md:21]: Should be `found here` instead of `found at here`.

File: README.md [hunk 2]
```diff
@@ -15,106 +15,73 @@ The `SetForeground(u32)` method can be called to change the active foreground pr
...
 
-RON configuration files are stored at the following locations:
+> The best-effort and realtime classes have priority levels between 0 and 7, where 7 has the least priority, and 0 is the highest priority
 
-- User-config: `/etc/system76-scheduler/exceptions/`
-- Distribution: `/usr/share/system76-scheduler/exceptions/`
+### Assignments
 
-The files contain a list of process names that are prohibited from having priority adjustments.
+Children of the profile define what processes to assign to it
```
Comment (jacobgkau) [inline README.md:66]: This sentence is missing punctuation.

**Review by jacobgkau — CHANGES_REQUESTED** [review]
I'm testing this out, but noticed a couple of small things while looking over the README.

**Comment by jacobgkau** [comment]
If I run `nice -n 12 stress -c 1`, the process starts with niceness 12. If I switch focus away from the terminal (and wait a bit), the niceness goes down to 6; switching back to the terminal then causes it to go down to 0.

Screenshot from 2023-03-09 16-52-40

I don't think this is actually a regression, but are manual nice values supposed to work?

**Comment by jacobgkau** [comment]
Confirmed `nice` not sticking isn't a regression (it didn't work to begin with before, so it's closer to working now.)

@mmstick If I comment out the background/foreground assignments in `config.kdl`, then none of the other assignments seem to apply anymore. Can you confirm if there's some other way that a user should turn off the background/foreground priority switching if they still want to use manual assignments? (This has come up in previous versions as a desired use case, and currently works.)

**Comment by jacobgkau** [comment]
It occurred to me that since exceptions aren't working anyway, maybe setting foreground and background to both have the same neutral settings would be the workaround to "not use" that part of the scheduler. If that's the case, then we might want to consider moving those two assignments to the top of the config file and adding an explanation to the comments that they need to be set to something for the rest to work.

**Comment by mmstick** [comment]
@jacobgkau I've been working on a solution for handling priority for exceptions, and the latest push should resolve that. The scheduler can now detect if a process was exec'd from a nice command and ignore priority assignments.

I've also fixed the issue with foreground profiles interfering with other profiles. Disabling them will no longer cause other profile to also be disabled.

**Comment by jacobgkau** [comment]
On 330d0c3, I'm seeing processes only get the foreground priority and never the background priority, whether they're focused or not. (This happens even after deleting the `/etc/system76-scheduler/` directory; if I open e.g. Nautilus and then switch focus to Terminal, Nautilus still keeps the default foreground niceness of 0.)

`nice` values set by launching with `nice -n ...` do stick now, so that seems to be fixed.

While testing, I noticed Kdenlive (installed via Flatpak) gets a niceness of `-3`, which seems to be the configured niceness for `desktop-environment` in the default config?

**Comment by jacobgkau** [comment]
Seems it might be any Flatpak app getting classified as `desktop-environment`? LibreWolf and VS Code installed via Flatpak have similar issues.

For VS Code, that app spawns multiple `code` processes. I see one that gets `-3` (`desktop-environment`), several that get `0` (foreground), and one that gets `9` (`session-services`).

**Comment by mmstick** [comment]
Flatpak process are now fixed. The issue was that the parent process (bwrap) was not known to the scheduler at the time it was notified of the process being created. The scheduler will now attempt to add parent processes if they're missing on adding a new process.

Foreground process management has also been refactored . The old code was not a good fit for the current logic, since processes are now managed internally in a different way.

Although foreground process detection can break if the service is being restarted at the same time pop-shell is trying to send a new foreground PID to it.

**Comment by jacobgkau** [comment]
I thought I saw foreground/background detection working again on 300e980, but now I'm not seeing it work anymore (tried rebooting before and after removing my `/etc` config, still nothing.) I might be running into the case you're aware of where it can break.

If I disable foreground/background detection (by commenting those lines out), at least some of the other assignments are still taking place now. 

When background/foreground are commented out, even our deb-packaged Firefox gets `-3` (`desktop-environment`) instead of a default `0` if I launch from its dock icon, but it gets `9` if I launch from the terminal (presumably inherited from `gnome-terminal-server` which is classified as `session-services`.) Is this how it's supposed to work when foreground/background don't have assignments?

A custom assignment group called `custom` (in `/etc/system76-scheduler/assignments/custom.kdl`) containing `stress` and assigned a different nice value in `config.kdl` does not seem to apply to a plain `stress -c 1` process anymore (I thought this was working earlier in the PR.)

**Comment by mmstick** [comment]
@jacobgkau The new change to the config makes it possible to add exclude rules to a profile, and the config now has an exclude that'll fix gedit and nautilus.

**Comment by jacobgkau** [comment]
It's taking more time than expected for foreground/background priorities to apply when I switch between e.g. Nautilus and Firefox, sometimes 3-5 seconds or longer. On the current stable system76-scheduler, these usually apply within a second.

**Comment by jacobgkau** [comment]
On 2546566, it's still taking 3 updates of `top` to reflect foreground niceness changes, versus 1 on the released version of system76-scheduler.

**Comment by mmstick** [comment]
Delays for foreground process assignments, and assignments of new processes, is now fixed. Changes will fix the main loop being put to sleep between process changes, and remembers assigned priorities so that it won't have to look up a compatible profile a second time.

**Review by jacobgkau — CHANGES_REQUESTED** [review]
Foreground changes take effect faster now, usually within a second. (I have still observed occasional instances where it takes more than one second.) Foreground/background detection is still working for .deb and Flatpak apps.

Running `stress -c 1` sometimes starts the niceness at 9 (it's been the first time after booting both times I've seen this happen), before changing to 0 or 6 a few seconds later (so it seems like profiles are still conflicting sometimes.) As another example, I've seen Firefox have the `desktop-environment` niceness before foreground/background detection kicks in (just after logging in.)

I'm able to change the foreground/background nicenesses using `/etc` configuration. 

Trying to test with the `stress -c 1` command, I found that something is breaking and it's changing from the foreground niceness to a niceness of 0. I didn't figure out what was going on until I tried changing the foreground niceness to something else, and saw that the `stress` niceness still goes to 0 after a while, even when no profiles in my configuration are configured to 0. (The default foreground niceness being 0 made it seem like it was getting stuck on the foreground profile, when that wasn't the case.)

Attempting to define a custom group containing `stress` and assigning it a different niceness, I see `stress -c 1` start off with the foreground niceness, then correctly change to my custom-defined niceness, so that does seem to be working. (Am I supposed to have to add `stress` as an exception under the foreground profile for this to work without the delay?)

Starting a process with `nice -n # ...` is unreliable. For `nice -n 18 stress -c 1`, the process either gets the custom-profile niceness or, if I disable the custom profile, it _sometimes_ starts with 19 niceness (1 foreground + 18 from the command), but then changes to 0 and remains there.

**Comment by jacobgkau** [comment]
> stress's cgroup matches the session services profile because it's assigned by the `*.scope` rule. I think then that we either remove that cgroup rule or expect stress be treated like a service.

session-services should have a niceness of 9, right? I don't understand if this is supposed to explain why it was being set to 0.

**Comment by mmstick** [comment]
@jacobgkau With the recent change, are you still seeing issues with stress and nice?

**Comment by jacobgkau** [comment]
Just talking about stress (not nice), with the default configuration, I'm seeing it always get `-1` now, whether the terminal window is focused or not. `top`, on the other hand, switches between `-1` and `6` based on if any terminal window is focused or not. Is this test case valid?

**Comment by mmstick** [comment]
@jacobgkau The issue with stress's dispatched child processes was that execsnoop-bpfcc cannot see processes created from a fork that didn't also call `exec` after the fork. I've made some changes to look for missing descendants of processes when assigning a new process, or when setting the foreground process. So all of the children of the main stress process are getting their foreground and background assignments now.

I've also set a 10 second delay on service start to give system process enough time to get their cgroups assigned to them by systemd, and a fix to the priority assignment logic to avoid assigning a priority from a condition when the cgroup is not yet defined. So any processes lacking a cgroup will get their cgroups set on the next process map refresh.

**Comment by jacobgkau** [comment]
At first, the most recent commit seemed to be good. Focus detection with `stress` is working more reliably reliably now. A custom profile also applies (I see `stress -c 1` get the foreground or background niceness briefly, but the custom niceness applies after no more than 1 second.) The `nice` command with `stress` seems to apply on top of the foreground value (`-n 18` gave a niceness of `17`), and is at least behaving consistently now.

I still started seeing `stress -c 1` get set to niceness `0` (with no profiles configured for that niceness) eventually. The issue stopped when I reset my configuration (removed the `/etc` configuration, leaving only the default `/usr/share` configuration.) I replicated the issue by only creating the following file in `/etc/system76-scheduler/assignments/custom.kdl`:

```
assignments {
        custom {
                stress
        }
}
```

I do not need to assign the `custom` profile to anything in `config.kdl`; simply having the above file present causes `stress -c 1` to get set to a niceness of `0`. Perhaps this was user error if people aren't intended to create profiles that aren't assigned in the configuration, but my impression was that if I didn't actually configure a profile for an assignment group, then it wouldn't do anything. (I had been commenting out the `custom` profile in `config.kdl` to try and disable it for testing, but had not been commenting out the entire `assignments/custom.kdl` file.) Should I consider this expected behavior, or is this an issue with the config handling?

**Comment by mmstick** [comment]
So we should ignore any profiles that have no properties defined? By default, a profile that doesn't assign a nice will get a nice setting of 0.

**Comment by jacobgkau** [comment]
> Profiles can be created in assignment configs to supplement the assignments defined in the main config. They'll inherit the properties of a profile of the same name if it was defined in the main config. Otherwise the profile is unique to that config.

Maybe I'm not understanding the purpose of assignment configs. My impression based on the file/directory names was:

- Assignment configs in the `assignments` directory map process names to a group name.
- Profile configs in the `process-scheduler` block of `config.kdl` map group names (including groups defined here instead of in the `assignments` directory) to niceness, etc.

You're saying that profiles in the `assignments` directory (in files containing a top-level `assignments` block) are equivalent to ones in the `process-scheduler`/`assignments` block of `config.kdl`, they just get merged/inherited when their names match? Is it possible to define other properties like niceness in the `assignments` directory? What's the purpose of having most assignments defined separately from the main config file by default?

**Comment by jacobgkau** [comment]
> So we should ignore any profiles that have no properties defined? By default, a profile that doesn't assign a nice will get a nice setting of 0.

If I was attempting to make a process get a niceness of zero, my first instinct would be to create an assignment/profile and set it to 0 explicitly, not to create a profile and not set the niceness. I think this behavior should at least be documented, if we don't change it.

**Comment by mmstick** [comment]
That is correct. The same logic is used to parse the assignments and exceptions elements in both the assignment configs and the main config. The difference for profile assignments in the main config is that the `pipewire`, `foreground`, and `background` profiles have special meaning to the scheduler; and profiles defined in the main config are memorized so that assignment configs can inherit their values when they use those names.

The assignments config files are a secondary source to extend the main config for cases where you may not want to modify the main config. This way we could have distribution and desktop-specific assignment config files. Providing a default config specific to Pop!_OS, or a config specific to cosmic-de/cosmic-gnome.

I'm open to ideas for improving the config syntax though if you have any.

File: data/pop_os.kdl [hunk 3]
```diff
@@ -31,6 +31,11 @@ assignments {
 		Xorg
 	}
 
+	session-services {
+		include parent="gnome-session-binary"
+        include parent="gvfsd"
```
Comment (jacobgkau) [inline data/pop_os.kdl:36]: This line didn't get indented far enough.

**Review by jacobgkau — CHANGES_REQUESTED** [review]

**Comment by jacobgkau** [comment]
Regarding the config structure:

- Why do the additional `assignments {}` and `exceptions {}` files have their own `assignments` and `exceptions` directories, when we still have to specify `assignments {}`/`exceptions {}` in the individual files? (The default `assignments/pop_os.kdl` includes an `exceptions {}` block, does that mean there isn't really a difference between those directories?) Would a single `config.d` or `conf.d` directory to store both assignments and exceptions work just as well?
- Could we theoretically move `assignments {}` and `exceptions {}` out of `process-scheduler {}` and into the top level of `config.kdl`, since those items are top-level in the other files? (Instead of that, it feels more like the additional files should also specify `process-scheduler {}`, but that would be more work/typing when creating new files.)

Maybe both of these could be improved by keeping the current `config.kdl` structure but calling the additional directory something that corresponds with `process-scheduler`, like `process-scheduler.d` or `processes.d`. That might make it more intuitive that the top-level `assignments {}` and `exceptions {}` blocks in the standalone files are in fact part of the `process-scheduler {}` configuration block in the main config file (or at least correspond with that level of configuration.)

Thoughts/counterpoints?

**Comment by mmstick** [comment]
The `exceptions` directory no longer applies, but it's a good point that `assignments` isn't the right name for placing these config files. I'm fine with changing it to `process-scheduler`.

**Comment by jacobgkau** [comment]
- [X] Focus-based assignments are working with the default config (both .debs and Flatpaks).
- [X] Focus-based assignments work with custom values in `/etc`.
- [X] `nice -n #` works with the default config.
- [ ] A custom profile with no niceness defined still has foreground/background applied.
    - Only foreground niceness is applied in this situation.
- [X] A custom profile with a niceness defined inline (under `process-scheduler/`) works.
- [X] A custom profile with a niceness defined separately (in `config.kdl`) works.
- [X] Config reloading works with `system76-scheduler daemon reload`, `systemctl reload com.system76.Scheduler.service`, and `systemctl restart com.system76.Scheduler.service`.
- [X] PipeWire integration works; when an app starts playing audio, it's assigned the niceness of the `pipewire` group and retains it until it's closed. It gets regular foreground/background niceness again when it's relaunched.
- [X] After disabling the foreground/background profiles, other profiles still work.

It seems like 54978313367222e810acd25dc13dda3c5e2a9743 is only partially working-- it does allow a custom group that doesn't have a niceness defined to get the foreground niceness, but it doesn't allow that group to get the background niceness when it's not focused. I don't consider this blocking now that I know it's a configuration error, but wanted to point it out.

I also noticed if I define a niceness inline under `process-scheduler/` and also in `config.kdl`, the one in `process-scheduler/` takes precedence. I suppose this is expected behavior if the purpose of `process-scheduler/` is to allow extending `config.kdl`.

Regarding PipeWire integration, I noticed that when I start playing audio in Firefox, `firefox-bin` gets the `pipewire` group niceness, but `Isolated Web Content` still gets background/foreground niceness. I was able to recreate audio stuttering on a 4-core/8-thread machine by having Firefox play a YouTube video in one workspace and then focusing a `stress -c 8` terminal window in another workspace. However, I was still able to recreate some stutter after manually using `renice -6` on `Isolated Web Content` (it may have been better, but only slightly.)

I tried to test this in Bitwig Studio and found that neither `BitwigStudio`, `bitwig-studio`, or `BitwigAudioEngine` were detected as part of the `pipewire` group even when I started playing audio (`BigwigAudioEngine` only exists while playing audio, so it's the one I expected to be detected.) I tested with Bitwig set to its PipeWire and PulseAudio options with no difference in behavior.

The feature does work as expected with Resolve, Totem, and MPV (all with a single process). Does this sound like PipeWire integration is working as expected, or do you want to look into inheritance and/or detection any more? (I'm not sure what the limitations are on the PipeWire side for reporting those processes.)

**Comment by mmstick** [comment]
Pipewire gives only the PID of the process that's directly connected to it, which is typically the same process that's decoding and playing the audio.

It sounds like Firefox is connected to Pipewire from its main process, but the actual decoding is happening in a background process. It should be resolvable with inheritance.

I'm not sure about BitWig at the moment. I'll have to try that out.

**Review by jacobgkau — APPROVED** [review]
- [X] Focus-based assignments are working with the default config (both .debs and Flatpaks).
- [X] Focus-based assignments work with custom values in `/etc`.
- [X] `nice -n #` works with the default config.
- [ ] A custom profile with no niceness defined still has foreground/background applied.
    - Only foreground niceness is applied in this situation. (Still an improvement over the current release, not a regression.)
- [X] A custom profile with a niceness defined inline (under `process-scheduler/`) works.
- [X] A custom profile with a niceness defined separately (in `config.kdl`) works.
- [X] Config reloading works with `system76-scheduler daemon reload`, `systemctl reload com.system76.Scheduler.service`, and `systemctl restart com.system76.Scheduler.service`.
- [X] PipeWire integration works; when an app starts playing audio, it's assigned the niceness of the `pipewire` group and retains it until it's closed. It gets regular foreground/background niceness again when it's relaunched.
- [X] After disabling the foreground/background profiles, other profiles still work.

The `pipewire` profile niceness now applies to child processes (e.g. `Isolated Web Content` in addition to `firefox-bin`.) Bitwig still doesn't get this profile even after it's playing audio for some reason. Kdenlive also doesn't get it. (These are both apps that you'd typically have focused when they're playing audio, though.) It still works with MPV, VLC, Totem, Resolve, and Elisa.

Everything is working as well or better than the current release at this point. I'm comfortable approving this and PipeWire detection can be made more comprehensive later if possible, or we can keep working on it now if it's easier.

**Comment by mmstick** [comment]
I'm fine with releasing as is. It's a substantial improvement and these remaining flaws are easily fixed in a backwards-compatible way.

**Review by 13r0ck — COMMENTED** [review]
I guess if nobody else is going to review this I can

I don't see any issues with the code, I have a few small requests, but I can approve if you disagree

1. Because of execsnoop's env variable, the build isn't just `cargo build` like most rust projects are, so it would be nice to have a not in the README.md about using just to build from source
2. Similar to (1) a `just clippy` that builds and checks for clippy lints would be a nice to have
3. As this is a rewrite, I think now is a good time to bring up adding `#![deny(missing_docs)]` to this project just to make sure everything is documented well

**Comment by 13r0ck** [comment]
If I say please, can we get `#![deny(missing_docs)]` added to the crates here?

**Comment by mmstick** [comment]
You mean to execsnoop?

**Comment by 13r0ck** [comment]
The daemon would be nice too, but as that isnt a library I don't think it is necessary

**Comment by mmstick** [comment]
execsnoop would be the only component that's not directly tied to the daemon

**Comment by 13r0ck** [comment]
Yep for sure, I think documentation would be nice for if someone else (internal or community) wants to make a PR.

**Pull request merged**